### PR TITLE
Fix 65fef82 (split of region bigger than vklimits)

### PR DESCRIPTION
**Pull request by rjodinchr**
Looking for region_offset and region_group_offset to know if we can
split the region is not enough as a test using only one of them can
still be split. But it will not be accepted by clvk with the current
test.
Instead let's looking into the kernel options to choose whether we try
it or not.

Also round_up the size of the workgroup to the local size in order to
keep region uniform.

### CODE REVIEW

File: src/program.hpp [hunk 1]
```diff
@@ -556,9 +557,19 @@ struct cvk_program : public _cl_program, api_object<object_magic::program> {
...
     }
 
-    bool has_region_offset() const {
-        return push_constant(pushconstant::region_offset) &&
-               push_constant(pushconstant::region_group_offset);
+    bool options_allow_split_region(std::string options) {
+        return options.find("-cl-std=CL2.0") != std::string::npos ||
+               options.find("-cl-std=CL3.0") != std::string::npos ||
+               options.find("cl-arm-non-uniform-work-group-size") !=
+                   std::string::npos;
```
Comment (kpet) [inline src/program.hpp:564]: This doesn't exactly mirror the set of conditions in clspv. It doesn't feel great to hardcode the same set of conditions here. Maybe there is a case to have clspv pass the information in the SPIR-V (i.e. this program was compiled with support for non-uniform NDRanges)? Maybe we should always enable support for non-uniform NDRanges instead of splitting opportunistically? We're definitely going to need more tests to cover all the cases as well :).

File: src/queue.cpp [hunk 2]
```diff
@@ -549,8 +549,8 @@ cl_int cvk_command_kernel::dispatch_uniform_region_iterate(
 
     size_t num_splitted_regions =
         ceil_div(num_workgroups[dim], vklimits.maxComputeWorkGroupCount[dim]);
-    size_t splitted_region_gws =
-        ceil_div(region.gws[dim], num_splitted_regions);
+    size_t splitted_region_gws = round_up(
```
Comment (kpet) [inline src/queue.cpp:552]: I haven't tried to verify this but it feels like gws calculations are getting more complex than they should be. Each region's gws will be a round multiple of the number of workgroups since we've already done non-uniform splitting so the gws calculation should be as simple as:
```
gws = N * lws
```

I suggest we refactor the logic to make this apparent as a separate PR.

**Review by kpet — COMMENTED** [review]
Getting a bit late here but I think I got this right. Let's see if tomorrow me still agrees :).

File: src/program.hpp (hunk 1 above)
Comment (rjodinchr) [reply src/program.hpp:564]: I feel like always enabling the non-uniform NRRanges is the right move here.
It will be like that anyway when the default version will be 3.0.

File: src/queue.cpp (hunk 2 above)
Comment (rjodinchr) [reply src/queue.cpp:552]: I think I got things too complicated because for some reason I wanted to have regions that are of the same size.
But it will definitely be easier to just make regions reach the limit, except for the last one.
```
size_t splitted_region_gws = vklimits.maxComputeWorkGroupCount[dim] * region.lws[dim];
```
We already have the `std::min` doing what we want to deal with the last region.

File: src/program.hpp (hunk 1 above)
Comment (kpet) [reply src/program.hpp]: Yes, we'll have that on by default once we switch to 3.0 (we'll need to rework the clspv flags so the only way to do this is not a deprecated Arm extension). There are ways to optimise further by having multiple variants of the compiled programs but we can treat that as an optimisation for later. Let's always enable non-uniform NDRanges for now. I'll capture options for further optimisations as part of merging this PR.

File: src/queue.cpp (hunk 2 above)
Comment (kpet) [reply src/queue.cpp]: Yup, easier to follow now, thanks!

**Review by kpet — CHANGES_REQUESTED** [review]
Let's get rid of all the complexity and always enable support for non-uniform NDRanges.

**Review by kpet — APPROVED** [review]
Thanks!
